src/sync.h

```diff
@@ -223,16 +242,77 @@ class SCOPED_LOCKABLE UniqueLock : public Base
      friend class reverse_lock;
 };
 
+template <typename Mutex, typename Base = typename Mutex::SharedLock>
+class SCOPED_LOCKABLE SharedLock : public Base
+{
+private:
+    void SharedEnter(const char* pszName, const char* pszFile, int nLine)
+    {
+        EnterCritical(pszName, pszFile, nLine, Base::mutex());
+#ifdef DEBUG_LOCKCONTENTION
+        if (!Base::try_lock()) {
+            PrintLockContention(pszName, pszFile, nLine);
+#endif
+            Base::lock();
+#ifdef DEBUG_LOCKCONTENTION
+        }
+#endif
+    }
+
+    bool TrySharedEnter(const char* pszName, const char* pszFile, int nLine)
+    {
+        EnterCritical(pszName, pszFile, nLine, Base::mutex(), true);
+        if (Base::try_lock()) {
+            return true;
+        }
+        LeaveCritical();
+        return false;
+    }
+
+public:
+    SharedLock(Mutex& mutexIn, const char* pszName, const char* pszFile, int nLine, bool fTry = false) SHARED_LOCK_FUNCTION(mutexIn) : Base(mutexIn, std::defer_lock)
+    {
+        if (fTry) {
+            TrySharedEnter(pszName, pszFile, nLine);
+        } else {
+            SharedEnter(pszName, pszFile, nLine);
+        }
+    }
+
+    SharedLock(Mutex* pmutexIn, const char* pszName, const char* pszFile, int nLine, bool fTry = false) SHARED_LOCK_FUNCTION(pmutexIn)
+    {
+        if (!pmutexIn) return;
+
+        *static_cast<Base*>(this) = Base(*pmutexIn, std::defer_lock);
+        if (fTry) {
+            TrySharedEnter(pszName, pszFile, nLine);
+        } else {
+            SharedEnter(pszName, pszFile, nLine);
+        }
+    }
```

Comment:
<s> why we need  even to consider a situation when `pmutexIn` can be non-null ptr? </s>
ok it seems as copy-paste of `UniqueLock`

```suggestion
```